File: src/examples/ruby/proxy/proxy_server.rb
# Copyright 2011 the original author or authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require "vertx"
include Vertx

client = HttpClient.new
client.port = 8282
client.host = "localhost"

HttpServer.new.request_handler do |req|
  puts "Proxying request: #{req.uri}"

  c_req = client.request(req.method, req.uri) do |c_res|
    puts "Proxying response: #{c_res.status_code}"
    req.response.chunked = true
    req.response.status_code = c_res.status_code
    req.response.headers.update(c_res.headers)
    c_res.data_handler do |data|
      puts "Proxying response body: #{data}"
      req.response.write_buffer(data);
    end
    c_res.end_handler { req.response.end }
  end
  c_req.chunked = true
  c_req.headers.update(req.headers)
  req.data_handler do |data|
    puts "Proxying request body #{data}"
    c_req.write_buffer(data)
  end
  req.end_handler { c_req.end }

end.listen(8080)

File: src/examples/javascript/proxy/proxy_server.js
/*
 * Copyright 2011-2012 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
load('vertx.js')

var client = vertx.createHttpClient().setPort(8282);

vertx.createHttpServer().requestHandler(function(req) {
  stdout.println("Proxying request: " + req.uri);

  var c_req = client.request(req.method, req.uri, function(c_res) {
    stdout.println("Proxying response: " + c_res.statusCode);
    req.response.statusCode = c_res.statusCode;
    req.response.setChunked(true);
    req.response.putAllHeaders(c_res.headers());
    c_res.dataHandler(function(data) {
      stdout.println("Proxying response body: " + data);
      req.response.writeBuffer(data);
    });
    c_res.endHandler(function() { req.response.end() });
  });
  c_req.setChunked(true);
  c_req.putAllHeaders(req.headers());
  req.dataHandler(function(data) {
    stdout.println("Proxying request body " + data);
    c_req.writeBuffer(data);
  });
  req.endHandler(function() { c_req.end() });

}).listen(8080)